src/preferences/node/preference-server-module.ts

```diff
@@ -0,0 +1,80 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { ContainerModule, } from 'inversify';
+import { ConnectionHandler } from '../../messaging/common';
+import { JsonRpcProxyFactory } from '../../messaging/common/proxy-factory';
+import { IPreferenceServer } from '../common/preference-protocol'
+import { DefaultPreferenceServer } from '../node/default-preference-server'
+import { getRootDir } from '../../filesystem/node/filesystem-server-module'
+import { CompoundPreferenceServer } from '../common/compound-preference-server'
+import { JsonPreferenceServer, PreferencePath } from './json-preference-server'
+import { IPreferenceClient } from '../common/preference-protocol'
+import * as path from 'path';
+
+
+export const preferenceServerModule = new ContainerModule(bind => {
+    const WorkspacePreferenceServer = Symbol('WorkspacePreferenceServer');
+    const UserPreferenceServer = Symbol('UserPreferenceServer');
+
+    bind(DefaultPreferenceServer).to(DefaultPreferenceServer).inSingletonScope();
+
+    bind(WorkspacePreferenceServer).to(JsonPreferenceServer).inSingletonScope();
+    bind(UserPreferenceServer).to(JsonPreferenceServer).inSingletonScope();
+
+    // bind(JsonPreferenceServer).to(WorkspacePreferenceServer).inSingletonScope();
+    // bind(PreferencePath).toConstantValue(path.join('.theia', 'prefs.json')).whenInjectedInto(WorkspacePreferenceServer.toString());
+    // const userHome = ".default"; // FIXME
+    // bind(PreferencePath).toConstantValue(path.join(userHome, '.theia', 'prefs.json')).whenInjectedInto(UserPreferenceServer.toString());
+    // const home = ctx.container.get(FileSystemNode);
+    const root = getRootDir();
+    if (root) {
+        bind(PreferencePath).toConstantValue(path.join(root, '.theia', 'prefs.json'));
+    }
+
+    bind(IPreferenceServer).toDynamicValue(ctx => {
+        const defaultServer = ctx.container.get(DefaultPreferenceServer);
+        const userServer = ctx.container.get<IPreferenceServer>(UserPreferenceServer);
+        const workspaceServer = ctx.container.get<IPreferenceServer>(WorkspacePreferenceServer);
+        return new CompoundPreferenceServer(defaultServer, userServer, workspaceServer);
+    }).inSingletonScope();
+
+    // bind(IPreferenceServer).to(JsonPreferenceServer).inSingletonScope();
+    // bind(PreferencePath).toConstantValue(".theia/prefs.json");
+
+    bind<ConnectionHandler>(ConnectionHandler).toDynamicValue(ctx => {
+        let clients: IPreferenceClient[] = []
+        // const prefServers = ctx.container.getAll(JsonPreferenceServer);
+        // const prefServers: Map<string, IPreferenceServer> = new Map<string, IPreferenceServer>();
+        const prefServers: JsonPreferenceServer[] = [];
+        prefServers.push(ctx.container.get<JsonPreferenceServer>(WorkspacePreferenceServer));
```

Comment:
Let's add `setClient` method to `IPreferenceServer`, and implement it in `CompoundPreferenceServer` by setting the given client on each server.